pkg/controllers/machine/controller.go

```diff
@@ -0,0 +1,201 @@
+	controllerutil.AddFinalizer(machine, v1alpha5.TerminationFinalizer)
+	if !equality.Semantic.DeepEqual(machine, stored) {
+		if err := c.kubeClient.Patch(ctx, machine, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+	}
+
+	stored = machine.DeepCopy()
+	var results []reconcile.Result
+	var errs error
+	for _, reconciler := range []interface {
+		Reconcile(context.Context, *v1alpha5.Machine) (reconcile.Result, error)
+	}{
+		c.launch,
+		c.registration,
+		c.initialization,
+		c.liveness,
+	} {
+		res, err := reconciler.Reconcile(ctx, machine)
+		errs = multierr.Append(errs, err)
+		results = append(results, res)
+	}
+	if !equality.Semantic.DeepEqual(stored, machine) {
+		statusCopy := machine.DeepCopy()
+		if err := c.kubeClient.Patch(ctx, machine, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(multierr.Append(errs, err))
+		}
+		if err := c.kubeClient.Status().Patch(ctx, statusCopy, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(multierr.Append(errs, err))
+		}
+	}
+	return result.Min(results...), errs
+}
+
+func (c *Controller) Finalize(ctx context.Context, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	stored := machine.DeepCopy()
+	if !controllerutil.ContainsFinalizer(machine, v1alpha5.TerminationFinalizer) {
+		return reconcile.Result{}, nil
+	}
+	if err := c.cleanupNodeForMachine(ctx, machine); err != nil {
+		if termination.IsNodeDrainError(err) {
+			return reconcile.Result{Requeue: true}, nil
+		}
+		return reconcile.Result{}, nil
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("provider-id", machine.Status.ProviderID))
+	if err := c.cloudProvider.Delete(ctx, machine); cloudprovider.IgnoreMachineNotFoundError(err) != nil {
+		return reconcile.Result{}, fmt.Errorf("terminating cloudprovider instance, %w", err)
+	}
+	controllerutil.RemoveFinalizer(machine, v1alpha5.TerminationFinalizer)
+	if !equality.Semantic.DeepEqual(stored, machine) {
+		if err := c.kubeClient.Patch(ctx, machine, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(fmt.Errorf("removing machine termination finalizer, %w", err))
+		}
+		logging.FromContext(ctx).Infof("deleted machine")
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(ctx context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(controllerruntime.
+		NewControllerManagedBy(m).
+		For(&v1alpha5.Machine{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Watches(
+			&source.Kind{Type: &v1.Node{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
+				node := o.(*v1.Node)
+				machineList := &v1alpha5.MachineList{}
+				if err := c.kubeClient.List(ctx, machineList, client.MatchingFields{"status.providerID": node.Spec.ProviderID}, client.Limit(1)); err != nil {
```

Comment:
what would happen if there were two machines with the same status.providerID?

Reply:
We mark the machine as unhealthy and then should trigger the deletion flow for the machine. This breaks a _lot_ of assumptions that we have about providerID uniqueness. We could alternatively just decide to bail and crash if we get two nodes with the same providerID since this also breaks how we are tracking nodes in cluster state.